tools/empath-split.py

```diff
@@ -0,0 +1,233 @@
+      addr = func.offset + func.size - 1
+      while addr > func.offset:
+        loc = sm.lookup(addr, func.offset)
+        # This means there is no source map mappings for the entire function
+        # (because we give func.offset as a lower bound). Exit the loop.
+        if not loc:
+          break
+        # Exit the loop only if a location info with source file information is
+        # found. If not, continue the search.
+        if loc.source:
+          break
+        addr -= 1
+
+      if loc and loc.source:
+        func_to_src[func_name] = normalize_path(loc.source)
+      else:
+        if not is_synthesized_func(func_name):
+          diagnostics.warn(f"No source file information found in the source map for function '{func_name}'")
+
+    for func_name, src in func_to_src.items():
+      if src not in src_to_funcs:
+        src_to_funcs[src] = []
+      src_to_funcs[src].append(func_name)
+
+  # Visit paths in the reverse sorting order, so that we can process inner paths
+  # first.
+  # e.g. If we have /a/b and /a/b/c, /a/b/c will come first, so we can assign
+  # functions contained in /a/b/c to it first and assign the maining functions
+  # to /a/b.
+  visited_funcs = set()
+  path_to_funcs = {}
+  for path in sorted(paths, reverse=True):
+    ppath = PurePath(path)
+    path_to_funcs[path] = []
+    for src, funcs in src_to_funcs.items():
+      psrc = PurePath(src)
+      if ppath == psrc or ppath in psrc.parents:
+        for func in funcs:
+          if func not in visited_funcs:
+            visited_funcs.add(func)
+            path_to_funcs[path].append(func)
+  return path_to_funcs
+
+
+def main():
+  args, forwarded_args = parse_args()
+  wasm = os.path.expanduser(args.wasm)
+  paths_file = os.path.expanduser(args.paths_file)
+  sourcemap = os.path.expanduser(args.sourcemap) if args.sourcemap else None
+  if args.wasm_split:
+    wasm_split = os.path.expanduser(args.wasm_split)
+  else:
+    wasm_split = os.path.join(config.BINARYEN_ROOT, 'bin', 'wasm-split')
+
+  if not os.path.isfile(wasm):
+    exit_with_error(f"'{wasm}' was not found or not a file")
+  if not os.path.isfile(paths_file):
+    exit_with_error(f"'{paths_file}' was not found or not a file")
+  if sourcemap:
+    if not os.path.isfile(sourcemap):
+      exit_with_error(f"'{sourcemap}' was not found or not a file")
+  if not os.path.isfile(wasm_split):
+    exit_with_error(f"'{wasm_split}' was not found or not a file")
+
+  with open(paths_file, encoding='utf-8') as f:
+    paths = [normalize_path(path.strip()) for path in f if path.strip()]
+    # To make /a/b/c and /a/b/c/ equivalent
+    paths = [path.rstrip(os.sep) for path in paths]
+    # Remove duplicates
+    paths = list(dict.fromkeys(paths))
+
+  # Compute {path: list of functions} map
+  path_to_funcs = get_path_to_functions_map(wasm, sourcemap, paths)
+
+  # Write .manifest file
+  f = tempfile.NamedTemporaryFile(suffix=".manifest", mode='w+', delete=False)
```

Comment:
Can you avoid the `try/catch` here and instead use `delete_on_close=False, delete=args.preserve_manifest`?

Then do `with tempfile.NamedTemporaryFile(..) as f:`?

Reply:
Done. I don't think I need `delete_on_close=False` because I replaced `f.close()` with `f.flush()`. (I did `f.close()` to make sure the file is written before being used but `f.flush()` can do the same thing)